tlspuffin/src/tls/fn_utils.rs

```diff
@@ -50,34 +50,56 @@ pub fn fn_append_transcript(
     Ok(new_transcript)
 }
 
-pub fn fn_decrypt_handshake(
-    application_data: &Message,
+pub fn fn_new_flight() -> Result<MessageFlight<Message, OpaqueMessage>, FnError> {
+    Ok(MessageFlight::new())
+}
+
+pub fn fn_append_flight(
+    flight: &MessageFlight<Message, OpaqueMessage>,
+    msg: &Message,
+) -> Result<MessageFlight<Message, OpaqueMessage>, FnError> {
+    let mut new_flight = flight.clone();
+    new_flight.messages.push(msg.clone());
+    Ok(new_flight)
+}
+
+/// Decrypt a whole flight of handshake messages and return a Vec of decrypted messages
+pub fn fn_decrypt_handshake_flight(
+    flight: &MessageFlight<Message, OpaqueMessage>,
     server_hello_transcript: &HandshakeHash,
     server_key_share: &Option<Vec<u8>>,
     psk: &Option<Vec<u8>>,
     group: &NamedGroup,
     client: &bool,
     sequence: &u64,
-) -> Result<Message, FnError> {
-    let (suite, key, _) = tls13_handshake_traffic_secret(
-        server_hello_transcript,
-        server_key_share,
-        psk,
-        !*client,
-        group,
-    )?;
-    let decrypter = suite
-        .tls13()
-        .ok_or_else(|| FnError::Crypto("No tls 1.3 suite".to_owned()))?
-        .derive_decrypter(&key);
-    let message = decrypter
-        .decrypt(
-            PlainMessage::from(application_data.clone()).into_unencrypted_opaque(),
-            *sequence,
-        )
-        .map_err(|_err| FnError::Crypto("Failed to decrypt it fn_decrypt_handshake".to_string()))?;
-    Message::try_from(message)
-        .map_err(|_err| FnError::Crypto("Failed to create Message from decrypted data".to_string()))
+) -> Result<Vec<Message>, FnError> {
+    let mut sequence_number = *sequence;
+
+    let mut decrypted_flight = vec![];
+
+    for msg in &flight.messages {
+        if let MessagePayload::ApplicationData(_) = &msg.payload {
+            let mut decrypted_msg = fn_decrypt_multiple_handshake_messages(
+                &msg,
+                server_hello_transcript,
+                server_key_share,
+                psk,
+                group,
+                client,
+                &sequence_number,
+            )?;
+
+            decrypted_flight.append(&mut decrypted_msg);
+            sequence_number += 1;
+        }
+    }
+
+    println!("Decrypting handshake");
+    for msg in &decrypted_flight {
+        println!("DECRYPTED: {:?}", msg);
```

Comment:
same here

Reply:
actually remove that for loop

Reply:
yes I already removed that in another commit in the new-queries branch, I forgot some debug prints